docs/workflows/best-practices-osh.md

```diff
@@ -0,0 +1,398 @@
+- See also the License Risk Evaluation Framework slide at the bottom of the wiki page: [OSH interpret results wiki page: https://softwareimprovementgroup.atlassian.net/wiki/spaces/SSM/pages/50317951024/How+to+interpret+Open-Source+health+findings](https://workflowy.com/#/c9c4745b6b6b)
+- Note that the risk depends on the context: 
+  - e.g. when developing open-source software, more licenses are acceptable.
+  - it also depends on how you use a library: it is sometimes needed/a solution to wrap the library in an executable component that can be used by calling, instead of becoming a part of the codebase.
+
+
+### 7. Handling detected lack of freshness [ROUGH DRAFT]
+- The teams are responsible for keeping libraries reasonably up to date: 
+  - small library updates can be updated as part of regular maintenance; 
+  - larger updates (e.g. major versions or frameworks) should be planned explicitly as part of a development sprint.
+
+### 8. _Handling detected lack of activity_ (TBD)
+
+## Handling your libraries
+
+
+### 9. Updating a library [ROUGH DRAFT]
+There can be several reasons to consider updating a library.
+
+Checklist before updating:
+- Does it have open issues?​
+- Is the license (still) compatible?​
+  - → [Handling detected license issues](#handling-detected-license-issues)
+- Does it contain optimizations for performance, maintainability, functionality?​
+  - ? not sure why this is exactly asked?
+- Is the new version compatible?  ​
+- How mature is the version?​
+- Is it stable enough?​
+
+Always Manage libraries with a package manager (don’t mix third party code with first party code)​
+
+- compatibility break
+  - Compatibility breaks can be caused by the framework/library no longer supporting the technologies used in the system.​
+  - Analyze the impact and consider the following things:​
+  - What is the impact of updating the system to make it compatible?​
+  - Are there alternatives of the library?​
+  - Are the technologies used still relevant for the future of the system?​
+
+- when is postponing updates a good idea?
+  - The longer you postpone updating, the bigger the eventual pain. As your system grows and evolves, the costs and risks of upgrading an old library increase. Such an accumulation of maintenance debt may lead to a much larger effort than in the case of smaller incremental updates.​
+  - For core functionalities, it is a good idea to stay on a stable version of the library. ​Being an early adopter of a new version comes with some risks. Bugs are common in immature versions of libraries as they are not thoroughly tested. Although you have the opportunity to contribute, by testing and potentially providing solutions to bugs you encounter in open source libraries, it might be a better practice to wait until the library becomes more mature.​​
```

Comment:
My experience is a bit different. At least in the Java ecosystem, popular libraries are fairly mature so if a new, stable version comes out it typically works. The typical workflow is that once a new major version comes out, you start testing it to assess the impact. Meanwhile, other people do the same and any serious problem that surfaces receives a patch-level fix. So after a couple of weeks, you know where you stand. This is typically one of
- Go ahead and update
- There's a compatibility break which requires fixing on your end, so plan the work
- There's an acknowledged compatibility break which requires fixing on their end. Wait and keep an eye on the releases.

Bottom line: if a new, stable version comes out: don't wait, start testing. If it's really core and really important, already start testing with release candidates.

Reply:
makes sense: also here, I do not remember the source of this (plausable deniability ;-) )